site-src/geps/gep-1058.md

```diff
@@ -0,0 +1,623 @@
+#### Child 1 - store route
+kind: HTTPRoute
+metadata:
+  name: store-route
+  namespace: store-ns
+spec:
+  parentRefs:
+    kind: HTTPRoute
+    name: my-example
+    namespace: prod-gw
+    sectionName: store
+...
+```
+
+Here are the changes and related rules:
+
+1. Support for Parent/child Route relationships is an "Exteneded" feature in the
+   Gateway API specification
+2. Add `AllowedRoutes` to [HTTPRouteRule][httpRouteRule] as optional
+    1. Default value is nil
+    2. A single `RouteRule` can contain `BackendRefs` OR `AllowedRoutes`, not both.
+    3. A parent `Route` may include `RouteRules` that use `BackendRefs` **AND**
+       `RouteRules` that use `AllowedRoutes`.
+    4. `AllowedRoutes` MUST be non-nil in a `RouteRule` to permit child `Routes`
+       to attach to that `RouteRule`.
+        1. Explanation: A parent `Route` will allow delegation to one or more
+           child `Routes` by specifying `AllowedRoutes` where `BackendRef` would
+           be used otherwise.
+    5. Multiple child routes may attach to a single `RouteRule`, as long as they
+       are distinct.
+3. Allow a Route’s `ParentRef` Kind to be the same type as the Route’s type
+    1. `Kind` is required when the `ParentRefs` is a route. `Kind` wil default to
+       `gateway` (current behavior)
+4. Add `Name` (type is [SectionName][sectionName]) to
+   [HTTPRouteRule][httpRouteRule] as optional
+    1. The `Name` field is expected to be added in the implementation of
+       [GEP-995 - Add Name to HTTPRouteRule and HTTPRouteMatch][gep-995]
+    2. Default value is nil
+    3. A child route may include a `SectionName` when its `ParentRef` is a route.
+        1. A child route “attaches" to the parent route’s `RouteRule` that matches
+           that `SectionName`. The child route must also match the criteria
+           specified in the `RouteRule’s AllowedRoutes`.
+    4. If the `ParentRefs` of a child route doesn’t include a `SectionName`, the
+       child route attaches to all of the Parent Route’s `RouteRules` having
+       `AllowedRoutes` criteria that the child route meets.
+5. A parent `Route` and any related child `Routes` must be the same type of
+   `Routes`.
+    1. Example: A child httpRoute can not point to a parent tcpRoute.
+6. A `Route` MUST NOT be a parent to a `Route` AND a child of another `Route`.
+    1. Explanation: This limitation avoids the need for specific
+       internal/terminal Route types, and the design of using a ParentRef from
+       the child to an AllowedRoutes slot on a parent allows a Route to know if
+       it is either a parent or child (or misconfigured as both) without
+       requiring the graph traversal that a design like ForwardTo may need to
+       avoid cycles.
+    2. If an implementation chooses to not enforce this limitation, that must be
+       documented along with how many tiers of parent/child route relationships
+       are allowed
+7. A Route’s `ParentRefs` can list more than one type of parent.
```

Comment:
I'd put this as "all parents of a Route must be of the same Group and Kind" if we want to do this. It certainly will make some implementation things easier.

Reply:
This could have some bearing on GAMMA, but I think it's manageable as long as we're careful/explicit in how routes will be composited together.

The case I'm thinking of is defining a canary HTTPRoute on the mesh for a service (which would have ParentRef mesh or service). I _think_ a viable way for this to composite safely, with expected behavior, would be for a supporting mesh-integrated Gateway that specifies an HTTPRoute with the mesh service as a backendRef to bascially parse the mesh HTTPRoute as a child route rather than expecting that route to explicitly bind to the Gateway too.